### PR TITLE
Reduce memory consumption in graph construction pipeline.

### PR DESCRIPTION
**Pull request by zheng-da**
*Description of changes:*
This PR uses a wrapper for numpy memmapped arrays and release the numpy array after it is created. This can reduce memory consumption at least from the perspective of virtual memory.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: python/graphstorm/gconstruct/utils.py [hunk 1]
```diff
@@ -302,6 +302,224 @@ def _get_arrs_out_dtype(arrs):
...
+    @property
+    def shape(self):
+        """ The shape of the array.
+        """
+        return self._shape
+
+    def astype(self, dtype):
+        """ Return an array with converted data type.
+        """
+        arr = copy.copy(self)
```
[the remark this answers is not in the document]
Comment (classicsong) [reply python/graphstorm/gconstruct/utils.py:489]: or self._dtype = dtype?

File: python/graphstorm/gconstruct/utils.py [hunk 2]
```diff
@@ -329,8 +547,8 @@ def _merge_arrs(arrs, tensor_path):
         for arr in arrs:
             out_arr[row_idx:(row_idx + arr.shape[0])] = arr[:]
             row_idx += arr.shape[0]
-        return out_arr
-    elif isinstance(arrs[0], HDF5Array):
+        return ExtNumpyWrapper(tensor_path, out_arr.shape, out_arr.dtype)
```
Comment (classicsong) [inline python/graphstorm/gconstruct/utils.py:550]: Why not pass out_arr to ExtNumpyWrapper?

File: python/graphstorm/gconstruct/utils.py [hunk 3]
```diff
@@ -329,8 +547,8 @@ def _merge_arrs(arrs, tensor_path):
...
+    elif isinstance(arrs[0], ExtMemArrayWrapper):
         arrs = [arr.to_numpy() for arr in arrs]
         return np.concatenate(arrs)
```
Comment (classicsong) [inline python/graphstorm/gconstruct/utils.py:553]: Does np.concatenate support concatenating memmaped arrays?

**Review by classicsong — COMMENTED** [review]
I don't understand why this PR can save memory consumption.

File: python/graphstorm/gconstruct/utils.py [hunk 4]
```diff
@@ -302,6 +302,224 @@ def _get_arrs_out_dtype(arrs):
...
+            idx = idx.numpy()
+        # If the idx are sorted.
+        if np.all(idx[1:] - idx[:-1] > 0):
+            arr = self._arr[idx]
+        else:
+            # There are two cases here: 1) there are duplicated IDs,
+            # 2) the IDs are not sorted. Unique can return unique
+            # IDs in the ascending order that meets the requirement of
+            # HDF5 indexing.
+            uniq_ids, reverse_idx = np.unique(idx, return_inverse=True)
```
[the remark this answers is not in the document]
Comment (zheng-da) [reply python/graphstorm/gconstruct/utils.py:404]: it probably will consume a lot of memory. we should fix it in other PRs.

File: python/graphstorm/gconstruct/utils.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (zheng-da) [reply python/graphstorm/gconstruct/utils.py:489]: what is the semantic of `astype`? I think this method shouldn't change the existing array.

File: python/graphstorm/gconstruct/utils.py (hunk 2 above)
Comment (zheng-da) [reply python/graphstorm/gconstruct/utils.py:550]: we want to release the original numpy array and open a new one when it is used.

File: python/graphstorm/gconstruct/utils.py (hunk 3 above)
Comment (zheng-da) [reply python/graphstorm/gconstruct/utils.py:553]: we have called to_numpy. it should return normal numpy arrays.

File: python/graphstorm/gconstruct/utils.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (classicsong) [reply python/graphstorm/gconstruct/utils.py:489]: Return a new array with the input dtype.
My question is will copy.copy make a copy of self._arr?

File: python/graphstorm/gconstruct/utils.py (hunk 3 above)
Comment (classicsong) [reply python/graphstorm/gconstruct/utils.py:554]: I c

**Review by classicsong — APPROVED** [review]
LGTM

**Pull request merged**